File: Python/InsertOperator.py
import sys

def calculate(a, b, op):
    return {'+': a+b, '-': a-b, '*': a*b, '/': a//b}[op]

def dfs(cnt):
    global select_op, trans_op, visited, number
    global minimum, maximum
    pre = '!'
    if cnt == N-1:
        interim = number[0]
        idx = 1
        sign = False
        for x in select_op:
            if x == '/':
                if interim < 0:
                    sign = True
                    interim = abs(interim)
            interim = calculate(interim, number[idx], x)
            if sign:
                interim = -interim
                sign = False
            idx += 1
        maximum = max(maximum, interim)
        minimum = min(minimum, interim)
        return

    for i in range(N-1):
        if not visited[i] and pre != trans_op[i]:
            select_op.append(trans_op[i])
            visited[i] = True
            dfs(cnt+1)
            pre = select_op.pop()
            visited[i] = False


#======main======
trans = ['+', '-', '*', '/']
N = int(input())
number = list(map(int, sys.stdin.readline().split()))
input_op = list(map(int, sys.stdin.readline().split()))
trans_op = []
select_op = []
visited = [False for i in range(N-1)]
minimum = 1000000000
maximum = -1000000000

for i in range(len(input_op)):
    for j in range(input_op[i]):
        trans_op.append(trans[i])

dfs(0)
print(maximum)
print(minimum)

File: Java/InsertOperator.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class InsertOperator {
    static int[] number;
    static int[] operator;

    static int maximum = -1000000000, minimum = 1000000000;

    public static void main(String[] arg) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");

        int numberCount = Integer.parseInt(st.nextToken());
        int operatorCount = 0;
        String[] inputNumber = br.readLine().split(" ");
        String[] inputOperator = br.readLine().split(" ");

        number = new int[inputNumber.length];
        operator = new int[inputOperator.length];
        for(int i=0; i<inputNumber.length; i++) number[i] = Integer.parseInt(inputNumber[i]);
        for(int i=0; i<inputOperator.length; i++) {
            operator[i] = Integer.parseInt(inputOperator[i]);
            operatorCount += operator[i];
        }
        Inserting(operatorCount, number[0], 1);

        System.out.println(maximum);
        System.out.println(minimum);
    }

    public static void Inserting(int operatorCnt, int total, int nextNumberIdx) {
        if(operatorCnt == 0) {
            maximum = Math.max(maximum, total);
            minimum = Math.min(minimum, total);
            return;
        }

        int nextTotal;
        for(int i=0; i<operator.length; i++) {
            if(operator[i] > 0) {
                nextTotal = Calculate(total, i, number[nextNumberIdx]);
                operator[i]--;
                Inserting(operatorCnt-1, nextTotal, nextNumberIdx+1);
                operator[i]++;
            }
        }
    }

    public static int Calculate(int curTotal, int operator, int nextNumber) {
        if(operator==0) {
            return curTotal+nextNumber;
        }
        else if(operator==1) {
            return curTotal-nextNumber;
        }
        else if(operator==2) {
            return curTotal*nextNumber;
        }
        else {
            return (curTotal < 0) ? (Math.abs(curTotal)/nextNumber)*-1 : curTotal/nextNumber;
        }
    }
}
